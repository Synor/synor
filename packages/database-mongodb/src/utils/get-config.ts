import { SynorError } from '@synor/core';
import { ConnectionString } from 'connection-string';

type MongoClientOptions = import('mongodb').MongoClientOptions;

type MongoDBDatabaseConfig = Required<Pick<MongoClientOptions, 'appname'>> & {
  database: string;
};

type MongoDBEngineConfig = {
  migrationRecordCollection: string;
};

const validProtocol = ['mongodb', 'mongodb+srv'];

export function getConfig(
  uri: string
): {
  databaseConfig: MongoDBDatabaseConfig;
  engineConfig: MongoDBEngineConfig;
} {
  try {
    const {
      protocol: parsedProtocol,
      hostname: host,
      path,
      params
    } = new ConnectionString(uri, {
      params: {
        appname: '@synor/database-mongodb',
        synor_migration_record_collection: 'synor_migration_record'
      }
    });

    if (!parsedProtocol) {
      throw new Error(`[URI] missing: protocol!`);
    }

    // `ConnectionString` replaces `+` in protocol with ` `
    const protocol = parsedProtocol.replace(' ', '+');

    if (!validProtocol.includes(protocol)) {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`);
    }

    if (!host) {
      throw new Error(`[URI] missing: host!`);
    }

    const database = path?.[0];

    if (!database) {
      throw new Error('[URI] missing: database!');
    }

    const databaseConfig: MongoDBDatabaseConfig = {
      appname: params!.appname,
      database
    };

    const engineConfig: MongoDBEngineConfig = {
      migrationRecordCollection: params!.synor_migration_record_collection
    };

    return {
      databaseConfig,
      engineConfig
    };
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', 'exception', error);
  }
}
