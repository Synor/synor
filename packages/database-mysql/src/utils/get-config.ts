import { SynorError } from '@synor/core'
import { ConnectionString } from 'connection-string'
import { readFileSync } from 'fs'
import { resolve as resolvePath } from 'path'

type ConnectionOptions = import('mysql2').ConnectionOptions
type TLSConnectionOptions = import('tls').ConnectionOptions

type SSLConfig = Pick<
  TLSConnectionOptions,
  'ca' | 'cert' | 'ciphers' | 'passphrase' | 'rejectUnauthorized'
> & {
  key?: string
}

type SSLParams = Pick<
  SSLConfig,
  'ciphers' | 'passphrase' | 'rejectUnauthorized'
> & {
  ca?: string
  key?: string
  cert?: string
}

type MySQLDatabaseConfig = Required<Pick<ConnectionOptions, 'database'>> &
  Pick<
    ConnectionOptions,
    'host' | 'port' | 'user' | 'password' | 'multipleStatements' | 'ssl'
  >

type MySQLEngineConfig = {
  migrationRecordTable: string
}

export function getConfig(
  uri: string
): {
  databaseConfig: MySQLDatabaseConfig
  engineConfig: MySQLEngineConfig
} {
  try {
    const {
      protocol,
      hostname: host,
      port,
      user,
      password,
      path,
      params,
    } = new ConnectionString(uri, {
      params: {
        synor_migration_record_table: 'synor_migration_record',
      },
    })

    if (!protocol) {
      throw new Error(`[URI] missing: protocol!`)
    }

    if (protocol !== 'mysql') {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`)
    }

    const database = path?.[0]

    if (!database) {
      throw new Error('[URI] missing: database!')
    }

    let ssl: string | SSLConfig | undefined

    const sslRaw: string | undefined = params!.ssl

    if (sslRaw) {
      try {
        const sslParams: SSLParams = JSON.parse(sslRaw)

        ssl = {
          ciphers: sslParams.ciphers,
          passphrase: sslParams.passphrase,
          rejectUnauthorized: sslParams.rejectUnauthorized,
        }

        if (sslParams.ca) {
          ssl.ca = readFileSync(resolvePath(sslParams.ca))
        }
        if (sslParams.cert) {
          ssl.cert = readFileSync(resolvePath(sslParams.cert))
        }
        if (sslParams.key) {
          ssl.key = readFileSync(resolvePath(sslParams.key)).toString()
        }
      } catch (_) {
        ssl = sslRaw
      }
    }

    const databaseConfig: MySQLDatabaseConfig = {
      database,
      host,
      port,
      user,
      password,
      multipleStatements: true,
      ssl,
    }

    const engineConfig: MySQLEngineConfig = {
      migrationRecordTable: params!.synor_migration_record_table,
    }

    return {
      databaseConfig,
      engineConfig,
    }
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', 'exception', error)
  }
}
