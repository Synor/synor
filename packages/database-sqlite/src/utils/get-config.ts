import { SynorError } from '@synor/core'
import { ConnectionString } from 'connection-string'
import Debug from 'debug'
import { resolve as resolvePath } from 'path'

type ConnectionConfig = import('better-sqlite3').Options

type SQLiteDatabaseConfig = {
  filepath: string
  fileMustExist: ConnectionConfig['fileMustExist']
  memory: ConnectionConfig['memory']
  readonly: ConnectionConfig['readonly']
  timeout: ConnectionConfig['timeout']
}

type SQLiteEngineConfig = {
  migrationRecordTable: string
  schema: string
}

const debug = Debug('@synor/database-sqlite')

const validHost = ['', '.']

export function getConfig(
  uri: string
): {
  databaseConfig: SQLiteDatabaseConfig
  engineConfig: SQLiteEngineConfig
} {
  try {
    const {
      protocol,
      hostname: host = '',
      path,
      params
    } = new ConnectionString(uri, {
      params: {
        schema: 'main',
        synor_migration_record_table: 'synor_migration_record',
        file_must_exist: false,
        memory: false,
        readonly: false,
        timeout: 5000
      }
    })

    if (!protocol) {
      throw new Error(`[URI] missing: protocol!`)
    }

    if (protocol !== 'sqlite') {
      throw new Error(`[URI] unsupported: protocol(${protocol})!`)
    }

    if (!validHost.includes(host)) {
      throw new Error(`[URI] unsupported: host(${host})!`)
    }

    let filepath: string

    if (!path) {
      debug('[URI] missing: path! Using in-memory database...')
      params!.memory = true
    }

    if (params!.memory) {
      filepath = ':memory:'
    } else {
      filepath = resolvePath(`${host}/${path!.join('/')}`)
    }

    const databaseConfig: SQLiteDatabaseConfig = {
      filepath,
      fileMustExist: JSON.parse(params!.file_must_exist),
      memory: JSON.parse(params!.memory),
      readonly: JSON.parse(params!.readonly),
      timeout: JSON.parse(params!.timeout)
    }

    const engineConfig: SQLiteEngineConfig = {
      migrationRecordTable: params!.synor_migration_record_table,
      schema: params!.schema
    }

    return {
      databaseConfig,
      engineConfig
    }
  } catch (error) {
    throw new SynorError('Invalid DatabaseURI', 'exception', error)
  }
}
