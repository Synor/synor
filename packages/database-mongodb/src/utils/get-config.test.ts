import { getConfig } from './get-config';

describe('utils:getConfig', () => {
  let uri: Parameters<typeof getConfig>[0];

  beforeEach(() => {
    uri = 'mongodb://mongo:mongo@127.0.0.1:27017/synor';
  });

  test('accepts mongodb uri', () => {
    expect(getConfig(uri)).toMatchSnapshot();
  });

  test('accepts mongodb+srv uri', () => {
    uri = uri.replace('mongodb:', 'mongodb+srv:');
    expect(getConfig(uri)).toMatchSnapshot();
  });

  test.each(['mongo:', 'mysql:'])(
    `throws if uri protocol is not valid`,
    protocol => {
      uri = uri.replace('mongodb:', protocol);
      expect(() => getConfig(uri)).toThrow();
    }
  );

  test('throws if protocol is missing', () => {
    uri = uri.replace('mongodb:', '');
    expect(() => getConfig(uri)).toThrow();
  });

  test('throws if host is missing', () => {
    uri = uri.replace('127.0.0.1', '');
    expect(() => getConfig(uri)).toThrow();
  });

  test('throws if database is missing', () => {
    uri = uri.replace('/synor', '');
    expect(() => getConfig(uri)).toThrow();
  });

  test('throws if uri is malformed', () => {
    uri = 'mongodb://@ _ @/synor';
    expect(() => getConfig(uri)).toThrow();
  });

  test('accepts custom migration record collection name', () => {
    const collectionName = 'migration_history';
    uri = `${uri}?synor_migration_record_collection=${collectionName}`;
    expect(getConfig(uri).engineConfig.migrationRecordCollection).toBe(
      collectionName
    );
  });
});
