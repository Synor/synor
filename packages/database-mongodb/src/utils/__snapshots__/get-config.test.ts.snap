// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils:getConfig accepts mongodb uri 1`] = `
Object {
  "databaseConfig": Object {
    "appname": "@synor/database-mongodb",
    "database": "synor",
  },
  "engineConfig": Object {
    "migrationRecordCollection": "synor_migration_record",
  },
}
`;

exports[`utils:getConfig accepts mongodb+srv uri 1`] = `
Object {
  "databaseConfig": Object {
    "appname": "@synor/database-mongodb",
    "database": "synor",
  },
  "engineConfig": Object {
    "migrationRecordCollection": "synor_migration_record",
  },
}
`;
